### ISSUE TITLE
Rules should consider infix functions

### ISSUE THREAD

**Issue by mklnwt** [pinterest/ktlint#2460]
For example `argument-list-wrapping` & `wrapping`.

## Expected Behavior
```Kotlin
val code =
    """
    val x = every {
        foo.bar(baz)
    } returns bar
    """.trimIndent()
val formattedCode =
    """
    val x = every {
        foo.bar(
            baz
        )
    } returns bar
    """.trimIndent()
argumentListWrappingRuleAssertThat(code)
    .withEditorConfigOverride(MAX_LINE_LENGTH_PROPERTY to 10)
    .isFormattedAs(formattedCode)
```

## Current Behavior
The example above fails. Below works though.

```Kotlin
val code =
    """
    val x = every {
        foo.bar(baz)
    }.returns(bar)
    """.trimIndent()
val formattedCode =
    """
    val x = every {
        foo.bar(
            baz
        )
    }.returns(bar)
    """.trimIndent()
argumentListWrappingRuleAssertThat(code)
    .withEditorConfigOverride(MAX_LINE_LENGTH_PROPERTY to 14)
    .isFormattedAs(formattedCode)
```

## Additional information
* Current version of ktlint: 1.1.0 (tested on `master` as well)

**Comment by paul-dingemans** [comment]
This seems to be a regression bug from the `0.50.0` release. It works with `0.49.0`:
```
$ ktlint-0.49.0 --relative -F --stdin
21:19:07.409 [main] INFO com.pinterest.ktlint.cli.internal.KtlintCommandLine - Enable default patterns [**/*.kt, **/*.kts]
    val x = every {
        foo.bar(baz)
    } returns bar

val x =
    every {
        foo.bar(
            baz,
        )
    } returns bar
<stdin>:1:1: Exceeded max line length (10) (standard:max-line-length)
<stdin>:1:5: Property name should use the screaming snake case notation when the value can not be changed (standard:property-naming)
<stdin>:1:6: Property name should use the screaming snake case notation when the value can not be changed (standard:property-naming)
<stdin>:1:9: Exceeded max line length (10) (standard:max-line-length)
<stdin>:2:1: Exceeded max line length (10) (standard:max-line-length)
<stdin>:2:18: Exceeded max line length (10) (standard:max-line-length)
```

**Comment by paul-dingemans** [comment]
It turns out that this is not a regression bug, but was a consicious decision to handle binary expressions for which at least one of the operands was a call expression having an argument list. As a rather simplistic measure the argument list wrapping was ignored whenever it was part of a binary expression. It resolved incorrect formatting of examples below:
Example:
```
// $MAX_LINE_LENGTH_MARKER         $EOL_CHAR
val foo1 = foobar(foo * bar) + "foo"
val foo2 = foobar("foo", foo * bar) + "foo"
val foo3 = foobar("fooo", foo * bar) + "foo"
```
If the argument-list-wrapping would not be ignored when inside a binary expression, the code above would be formatted as:
```
// Max line length marker:         #
val foo1 = foobar(foo * bar) + "foo"
val foo2 = foobar(
  "foo",
  foo * bar
) + "foo"
val foo3 = foobar(
  "fooo",
  foo * bar
) + "foo"
```

Although disguised, your example is also a `BINARY EXPRESSION`:
![Screenshot 2023-12-30 at 12 24 49](https://github.com/pinterest/ktlint/assets/5195292/1482fcd6-070e-42fa-bf37-8cc58a873a74)

The alternative in which the argument is wrapped, is a `DOT QUALIFIED EXPRESSION`:
![Screenshot 2023-12-30 at 12 25 33](https://github.com/pinterest/ktlint/assets/5195292/780a4e9c-35d9-47e2-af9d-5e4cd0ed7622)

The `binary-expression-wrapping` rule is now smarter.

It no longer has a rule dependency on the `value-argument-wrapping` rule. Instead, if a binary expression is used as value argument, and it causes the max line length to be exceeded, then it wraps the value argument.
```
            // $MAX_LINE_LENGTH_MARKER    $EOL_CHAR
            val foobar1 = Foo(1 * 2 * 3 * 4)
            val foobar2 = Foo(bar(1 * 2 * 3))
            val foobar3 = Foo(bar("bar" + "bazzzzzzzzzzz"))
            val foobar4 = Foo(bar("bar" + "bazzzzzzzzzzzz"))
```
is formatted as
```
            // $MAX_LINE_LENGTH_MARKER    $EOL_CHAR
            val foobar1 = Foo(
                1 * 2 * 3 * 4
            )
            val foobar2 = Foo(
                bar(1 * 2 * 3)
            )
            val foobar3 = Foo(
                bar(
                    "bar" + "bazzzzzzzzzzz"
                )
            )
            val foobar4 = Foo(
                bar(
                    "bar" +
                        "bazzzzzzzzzzzz"
                )
            )
```

If the left hand side of the binary expression is call expression with a lambda argument, then wrapping the lambda argument has priority over wrapping the other arguments of the call expression.
```
            // $MAX_LINE_LENGTH_MARKER       $EOL_CHAR
            fun foo() {
                require(bar != "barrrrrrrr") { "some longgggggggggggggggggg message" }
            }
```
is formatted as
```
            // $MAX_LINE_LENGTH_MARKER       $EOL_CHAR
            fun foo() {
                require(bar != "barrrrrrrr") {
                    "some longgggggggggggggggggg message"
                }
            }
```

If the binary expression contains an elvis operator, and does not fit on a single line, than the expression is wrapped before the elvis operator.
```
            // $MAX_LINE_LENGTH_MARKER                            $EOL_CHAR
            val foo = foobar ?: throw UnsupportedOperationException("foobar")
```
is formatted as
```
            // $MAX_LINE_LENGTH_MARKER                            $EOL_CHAR
            val foo =
                foobar
                    ?: throw UnsupportedOperationException(
                        "foobar"
                    )
```

**Comment by mklnwt** [comment]
Shouldn't it be wrapped after the elvis operator?

Other operators are wrapped after as well (e.g. `+`, `&&`).

**Comment by paul-dingemans** [comment]
> Shouldn't it be wrapped after the elvis operator?
> 
> Other operators are wrapped after as well (e.g. `+`, `&&`).

Reason for wrapping before the "elvis" is the alignment with chained methods:
```
foo
    .map { doSomething() }
   ?.let { somethingElse() }
   ?: otherwise()
```
Also the "elvis" can more or less be associated with the "else" keyword which is also aligned at beginning of line.
